Show guest difficulty author in beatmap overlay

cdwcgt opened the pull request:
which this looks like:
![image](https://user-images.githubusercontent.com/34775378/229356164-c8ef02e3-97ae-47a5-95aa-e22a9735cf71.png)
web:
![image](https://user-images.githubusercontent.com/34775378/229356212-52f529f1-befe-404c-931f-97570e0bcc8e.png)

cdwcgt committed:
show guest diff author in `BeatmapPicker`

cdwcgt committed:
make `guestMapperContainer` private

tests don't use it

peppy commented on osu.Game/Overlays/BeatmapSet/BeatmapPicker.cs:
```diff
@@ -88,6 +89,12 @@ public BeatmapPicker()
                                     Origin = Anchor.BottomLeft,
                                     Font = OsuFont.GetFont(size: 17, weight: FontWeight.Bold)
                                 },
+                                guestMapperContainer = new FillFlowContainer
```
This can probably be a `LinkFlowContainer` directly. `FillFlowContainer` with one child seems pointless.

peppy commented on osu.Game/Overlays/BeatmapSet/BeatmapPicker.cs:
```diff
@@ -198,11 +205,32 @@ private void updateDisplay()
             version.Text = beatmapInfo?.DifficultyName ?? string.Empty;
         }
 
+        private Drawable getGueatMapper(APIUser user)
```
gueat?

peppy commented on osu.Game/Overlays/BeatmapSet/BeatmapPicker.cs:
```diff
@@ -198,11 +205,32 @@ private void updateDisplay()
         {
+            guestMapperContainer.Clear();
+
+            if (beatmapInfo != null && beatmapSet?.Author.OnlineID != beatmapInfo.AuthorID)
+            {
+                if (BeatmapSet?.RelatedUsers?.Single(u => u.OnlineID == beatmapInfo.AuthorID) is APIUser user)
```
This is a mess of conditionals. Also you are accessing both `beatmapSet` and `BeatmapSet` (should match other usages in file and only use one of the two).

peppy commented on osu.Game/Overlays/BeatmapSet/BeatmapPicker.cs:
```diff
@@ -198,11 +205,32 @@ private void updateDisplay()
         {
+            guestMapperContainer.Clear();
+
+            if (beatmapInfo != null && beatmapSet?.Author.OnlineID != beatmapInfo.AuthorID)
+            {
+                if (BeatmapSet?.RelatedUsers?.Single(u => u.OnlineID == beatmapInfo.AuthorID) is APIUser user)
```
This can't be anything other than an `APIUser`. Shouldn't be casting.

cdwcgt replied:
In order to prevent it being `APIUser?` , I changed

peppy replied:
Sure, but `Single` is also weird. What happens if more than one user is returned by the API? It will crash the game.

cdwcgt replied:
only one `APIUser` will match the condition? because `beatmapInfo.AuthorID` can only match one user.
The only possible cases are null and one `APIUser` I think
But if this is the case, I think should use `FirstOrDefault`?

cdwcgt committed:
typo

cdwcgt committed:
nullable condition

cdwcgt committed:
use `LinkFlowContainer` directly

cdwcgt committed:
condition fix

cdwcgt committed:
useless using

Joehuu committed:
Fix more typos and adjust font size to match web

Joehuu commented on osu.Game/Overlays/BeatmapSet/BeatmapPicker.cs:
```diff
@@ -88,6 +89,14 @@ public BeatmapPicker()
+                                    AutoSizeAxes = Axes.Both,
+                                    Anchor = Anchor.BottomLeft,
+                                    Origin = Anchor.BottomLeft,
+                                    Margin = new MarginPadding { Bottom = 1 },
```
`starRatingContainer ` also applies this margin for alignment.

Joehuu reviewed (dismissed):
LGTM

Joehuu committed:
Use public `BeatmapSet` to match other usages

bdach commented on osu.Game/Overlays/BeatmapSet/BeatmapPicker.cs:
```diff
@@ -198,11 +207,29 @@ private void updateDisplay()
             version.Text = beatmapInfo?.DifficultyName ?? string.Empty;
         }
 
+        private void getGuestMapper(APIUser user)
+        {
+            guestMapperContainer.With(d =>
+            {
+                d.AddText("mapped by ");
+                d.AddUserLink(user);
+            });
+        }
```
For future reference, this is almost criminal to me. A `getGuestMapper()` method name suggests that it will return something and be a pure method. Instead, it is a `void` method, so it returns nothing, and it *also* has baffling side effects which are - even more bafflingly - applied via `.With()`, which is supposed to be syntactic sugar first and foremost.

I'll fix it myself to speed proceedings up, but just for future reference I guess.

bdach committed:
Inline disturbing `getGuestMapper()` method

Disturbing because name suggests pure method, but it is in fact `void`
and also performs side effects by about the most confusing means
possible.

bdach committed:
Remove usage of "diff" vernacular

bdach committed:
Merge branch 'master' into gd-onlinesetoverlay
